pkg/agent/proxy/proxier.go

```diff
@@ -583,23 +612,51 @@ func (p *proxier) installServices() {
 				// Remove LoadBalancer flows and configurations.
 				if len(toDelete) > 0 {
 					if err := p.uninstallLoadBalancerService(toDelete, uint16(pSvcInfo.Port()), pSvcInfo.OFProtocol); err != nil {
-						klog.ErrorS(err, "Failed to remove flows and configurations of Service", "Service", svcPortName)
+						klog.ErrorS(err, "Error when uninstalling LoadBalancer flows and configurations for Service", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr)
 						continue
 					}
 				}
 				// Install LoadBalancer flows and configurations.
 				if len(toAdd) > 0 {
-					if err := p.installLoadBalancerService(nGroupID, toAdd, uint16(svcInfo.Port()), svcInfo.OFProtocol, uint16(affinityTimeout), svcInfo.ExternalPolicyLocal()); err != nil {
-						klog.ErrorS(err, "Failed to install LoadBalancer flows and configurations of Service", "Service", svcPortName)
+					if err := p.installLoadBalancerService(groupID, toAdd, uint16(svcInfo.Port()), svcInfo.OFProtocol, uint16(affinityTimeout), svcInfo.ExternalPolicyLocal()); err != nil {
+						klog.ErrorS(err, "Error when installing LoadBalancer flows and configurations for Service", "Service", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr)
 						continue
 					}
 				}
 			}
 		}
 
 		p.serviceInstalledMap[svcPortName] = svcPort
-		p.addServiceByIP(svcInfo.String(), svcPortName)
+		p.addServiceByIP(svcInfoStr, svcPortName)
+	}
+}
+
+func compareEndpoints(endpointsCached map[string]k8sproxy.Endpoint, endpointsInstalled []k8sproxy.Endpoint) (map[string]k8sproxy.Endpoint, map[string]k8sproxy.Endpoint, []k8sproxy.Endpoint) {
+	// Map endpointsToCache is used to store the Endpoints actually installed.
+	endpointsToCache := map[string]k8sproxy.Endpoint{}
+	// Map endpointsToRemove is used to store the Endpoints that should be removed.
+	endpointsToRemove := map[string]k8sproxy.Endpoint{}
+	// Slice newEndpoints is used to store the Endpoints that are newly installed.
+	var newEndpoints []k8sproxy.Endpoint
```

Comment:
Yes, then we need to change this function
```go
func endpointKey(endpoint k8sproxy.Endpoint, protocol binding.Protocol) string {
	return fmt.Sprintf("%s/%s", endpoint.String(), protocol)
}
```
to 

```go
func endpointKey(endpoint string, protocol binding.Protocol) string {
	return fmt.Sprintf("%s/%s", endpoint, protocol)
}
```

Reply:
Since we need to update the reference counter of new Endpoints, then we need to get the key by calling `func endpointKey(endpoint k8sproxy.Endpoint, protocol binding.Protocol)` to get/set the reference counter.

Reply:
I didn't get your point, my thought is we could call function endpointKey() when constructing the `newEndpoints`. Then we could consume the key from the returned slice directly?

Reply:
Any difference in that way?

Reply:
We will calculate key by calling `endpointKey() ` eventually.